runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -273,10 +288,24 @@ public void onHeartbeatResponse(List<Windmill.ComputationHeartbeatResponse> resp
   @Override
   public void sendHealthCheck() {
     if (hasPendingRequests()) {
-      send(StreamingGetDataRequest.newBuilder().build());
+      send(HEALTH_CHECK_REQUEST);
     }
   }
 
+  @Override
+  protected void shutdownInternal() {
+    // Stream has been explicitly closed. Drain pending input streams and request batches.
+    // Future calls to send RPCs will fail.
+    pending.values().forEach(AppendableInputStream::cancel);
```

Comment:
add a check where we add to the map

Reply:
we do a lot of synchronization on `this` and cannot have it block shutdown since this will cause stuckness.  The retry logic for sends on stream restarts, keeping sends serial, etc depends on this synchronization.

I can add a `todo` to improve the locking mechanisms for AbstractWIndmillStream?

the shutdown signal is only propagated, how about we add a finally check after the send to check for shutdown and clear pending.  This is guaranteed to only be called once during the lifetime of the stream

Reply:
> I can add a todo to improve the locking mechanisms for AbstractWIndmillStream?

Since it is a problem introduced by the new code, I think we should fix it before merging. Bugs here will also affect cloud path, want to make sure we don't break anything on cloud path.

Can we guard all logic (IIUC, there shouldn't be any blocking IO) that depends on shutdown flag with the shutdown mutex? If we guard `pending.put` with the shutdown mutex, we can avoid this race by checking.

Reply:
added a shutdown lock to manage this